### PR TITLE
Add Nx.conjugate

### PR DESCRIPTION
**Pull request by Benjamin-Philip**
See #693.

### CODE REVIEW

**Comment by Benjamin-Philip** [comment]
@polvalente,

So far I have only added an implementation for `BinaryBackend`. I also haven't written implementations for both EXLA and Torchx before. Are there any pointers/suggestions to keep in mind when writing one?

**Comment by polvalente** [comment]
> 

We can hold off on Torchx and EXLA for now because they don't support Complex numbers yet. I have a WIP branch for Torchx

**Comment by Benjamin-Philip** [comment]
> We can hold off on Torchx and EXLA for now because they don't support Complex numbers yet. I have a WIP branch for Torchx

In that case do we merge, or do we close and wait for complex support in Torchx and EXLA?

File: nx/lib/nx.ex
Comment (polvalente) [inline nx/lib/nx.ex:5219]: There's an issue with the output type here.

Try a test/example in which the input is s64 ou f32. I'd expect the output type to be inferred incorrectly (it should be the corresponding complex type).

I think we need a `to_complex` function in Nx.Types, which promotes everything below f32 to c64, f64 -> c128 and keeps complex types as the same.

Then you would have somthing like:
```suggestion
    impl!(tensor).conjugate(%{tensor | type: Nx.Types.to_complex(tensor.type)}, tensor)
```

Comment (polvalente) [reply nx/lib/nx.ex:5219]: Regarding EXLA/Torchx, we can go ahead and merge this after this adjustment :)

Comment (polvalente) [reply nx/lib/nx.ex:5219]: Oh, and regarding the failing torchx tests, there's a list to which you can add the tests in torchx/test/nx_doctest_test.exs

Comment (Benjamin-Philip) [reply nx/lib/nx.ex:5219]: Ah, yes I forgot about it :)

File: nx/lib/nx.ex
Comment (polvalente) [inline nx/lib/nx.ex:5198]:
```suggestion
  If $z = a + bi = r e^\\theta$, $conjugate(z) = z^* = a - bi =  r e^{-\\theta}$
```

File: nx/lib/nx/type.ex
Comment (polvalente) [inline nx/lib/nx/type.ex:196]: We only need to promote f64 to c128, everything else fits in c64 if it's not already complex

**Review by polvalente — APPROVED** [review]

**Comment by josevalim** [comment]
Should we implement the gradient as we add these new functions?

**Pull request merged**
